From e15aead044af1807aa2f5d43346dd55c5f558d56 Mon Sep 17 00:00:00 2001
Message-Id: <e15aead044af1807aa2f5d43346dd55c5f558d56.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Arun R Murthy <arun.r.murthy@intel.com>
Date: Thu, 20 Nov 2014 12:44:51 +0530
Subject: [PATCH 63/75] MUST_REBASE [VPG]: adf/intel: Add support for plane
 attach and detach

A given plane can be attached to a given pipe if supported by the
hardware. Initializes pipe pointers in the overlay/plane structures

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: I122b8d3ebdbd210cf7e295ca7a2ffac3d7b712d4
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
---
 .../video/adf/intel/include/core/intel_dc_config.h |    3 +++
 drivers/video/adf/intel/intel_adf_device.c         |   26 +++++++++++++++++---
 2 files changed, 25 insertions(+), 4 deletions(-)

diff --git a/drivers/video/adf/intel/include/core/intel_dc_config.h b/drivers/video/adf/intel/include/core/intel_dc_config.h
index 983b26c..79761c7 100644
--- a/drivers/video/adf/intel/include/core/intel_dc_config.h
+++ b/drivers/video/adf/intel/include/core/intel_dc_config.h
@@ -180,6 +180,8 @@ struct intel_plane_ops {
 	/*FIXME: Have to put adf overlay ops*/
 	const struct adf_overlay_engine_ops adf_ops;
 #endif
+	int (*attach)(struct intel_plane *plane, struct intel_pipe *pipe);
+	int (*detach)(struct intel_plane *plane, struct intel_pipe *pipe);
 	int (*validate_custom_format)(struct intel_plane *plane, u32 format,
 		u32 w, u32 h);
 	int (*validate)(struct intel_plane *plane,
@@ -194,6 +196,7 @@ struct intel_plane_ops {
 
 struct intel_plane {
 	struct intel_dc_component base;
+	struct intel_pipe *pipe;
 	const struct intel_plane_capabilities *caps;
 	const struct intel_plane_ops *ops;
 };
diff --git a/drivers/video/adf/intel/intel_adf_device.c b/drivers/video/adf/intel/intel_adf_device.c
index fadb8f4..766ee42 100644
--- a/drivers/video/adf/intel/intel_adf_device.c
+++ b/drivers/video/adf/intel/intel_adf_device.c
@@ -144,22 +144,38 @@ static void post_obj_set_destroy(struct post_obj_set *set)
 	}
 }
 
-/*----------------------------------------------------------------------------*/
-/*
 static int intel_adf_device_attach(struct adf_device *dev,
 				struct adf_overlay_engine *eng,
 				struct adf_interface *intf)
 {
-	return 0;
+	struct intel_adf_overlay_engine *intel_eng = to_intel_eng(eng);
+	struct intel_plane *plane = intel_eng->plane;
+	struct intel_adf_interface *intel_intf = to_intel_intf(intf);
+	struct intel_pipe *pipe = intel_intf->pipe;
+
+	if (plane && plane->ops && plane->ops->attach)
+		return plane->ops->attach(plane, pipe);
+	else
+		return -EINVAL;
 }
 
 static int intel_adf_device_detach(struct adf_device *dev,
 				struct adf_overlay_engine *eng,
 				struct adf_interface *intf)
 {
-	return 0;
+	struct intel_adf_overlay_engine *intel_eng = to_intel_eng(eng);
+	struct intel_plane *plane = intel_eng->plane;
+	struct intel_adf_interface *intel_intf = to_intel_intf(intf);
+	struct intel_pipe *pipe = intel_intf->pipe;
+
+	if (plane && plane->ops && plane->ops->detach)
+		return plane->ops->detach(plane, pipe);
+	else
+		return -EINVAL;
 }
 
+/*----------------------------------------------------------------------------*/
+/*
 static int intel_adf_device_validate_custom_format(struct adf_device *dev,
 						struct adf_buffer *buf)
 {
@@ -587,6 +603,8 @@ static void intel_adf_device_state_free(struct adf_device *dev,
 
 static const struct adf_device_ops intel_adf_device_ops = {
 	.owner = THIS_MODULE,
+	.attach = intel_adf_device_attach,
+	.detach = intel_adf_device_detach,
 	.validate_custom_format = intel_adf_device_validate_custom_format,
 	.validate = intel_adf_device_validate,
 	.post = intel_adf_device_post,
-- 
1.7.9.5

